app/services/query.py

```diff
@@ -0,0 +1,220 @@
+                yield host
+
+    def get(self, service, ip_address):
+        ip_map = self.data.get(service)
+        if ip_map is None:
+            return None
+
+        host_dict = ip_map.get(ip_address)
+        if host_dict is None:
+            return None
+
+        host = host_dict.copy()
+        host['service'] = service
+        host['ip_address'] = ip_address
+        return host
+
+    def put(self, host):
+        service = host['service']
+        ip_address = host['ip_address']
+
+        ip_map = self.data.get(service)
+        if ip_map is None:
+            ip_map = {}
+            self.data[service] = ip_map
+
+        host_dict = host.copy()
+        del host_dict['service']
+        del host_dict['ip_address']
+
+        ip_map[ip_address] = host_dict
+
+    def delete(self, service, ip_address):
+        ip_map = self.data.get(service)
+        if ip_map is None:
+            return False
+
+        host_dict = ip_map.get(ip_address)
+        if host_dict is None:
+            return False
+
+        del ip_map[ip_address]
+        if len(ip_map) == 0:
+            del self.data[service]
+        return True
+
+
+class LocalFileQueryBackend(QueryBackend):
+    def __init__(self, file=tempfile.TemporaryFile().name):
+        self.backend = MemoryQueryBackend()
+        self.file = file
+        if os.path.isfile(self.file) and os.stat(self.file).st_size > 0:
+            self.backend.data = pickle.load(open(self.file))
+
+    def _save(self):
+        pickle.dump(self.backend.data, open(self.file, 'w'))
+
+    def query(self, service):
+        return self.backend.query(service)
+
+    def query_secondary_index(self, service_repo_name):
+        return self.backend.query_secondary_index(service_repo_name)
+
+    def get(self, service, ip_address):
+        return self.backend.get(service, ip_address)
+
+    def put(self, host):
+        self.backend.put(host)
+        self._save()
+
+    def delete(self, service, ip_address):
+        self.backend.delete(service, ip_address)
+        self._save()
+
+
+class DynamoQueryBackend(QueryBackend):
+    def query(self, service):
+        return self._read_cursor(Host.query(service))
+
+    def query_secondary_index(self, service_repo_name):
+        return self._read_cursor(Host.service_repo_name_index.query(service_repo_name))
+
+    def get(self, service, ip_address):
+        try:
+            host = Host.get(service, ip_address)
+            if host is None:
+                return None
+            return self._pynamo_host_to_dict(host)
+        except Host.DoesNotExist:
+            return None
+
+    def put(self, host):
+        self._dict_to_pynamo_host(host).save()
+
+    def batch_put(self, hosts):
+        with Host.batch_write() as batch:
+            for host in hosts:
+                batch.save(self._dict_to_pynamo_host(host))
+
+    # TODO is all of this pomp and circumstance really necessary to properly delete?
```

Comment:
one thing is to point out, now it does not throw Host.DoesNotExist but return None, can you check all of the callers if they are not broken after this change.